README.md
# NAME

Kikyo - Information Base for managing server assets

# DESCRIPTION

Kikyo is a Kikyo is a Information Base for managing server assets.

# HOW TO RUN

## mysql

First, create a database and a user on mysql-server. Then, create table with etc/schema.sql

## environment values

Set following environment values for database connection.

- KIKYO\_DSN - dns string for your database
- KIKYO\_DBUSER - database user
- KIKYO\_DBPASS - password for user authentication

Example.

    export KIKYO_DSN="dbi:mysql:host=10.0.0.1;database=kikyo"
    export KIKYO_DBUSER=kikyo
    export KIKYO_DBPASS=kiky0P@ss

Or, modify the app.psgi to conform to your database.

## plackup

Now, you can run Kikyo with plackup under Kikyo project directory.

# JSON API

## get racklist

    GET /v1/racklist

Returns list of rackname and host count.

## get rack info

    GET /v1/rack/{rackname}

Returns hosts list in specified rack.

## fetch host

    GET /v1/host/{hostid}

Returns specified host.

## search hosts

    GET /v1/search

Returns hosts list that matched for search query.

Available search parameters is following.

- id
- name
- rack
- ip
- hwid
- os
- modelname
- status

## add or update a host

    POST /v1/rack/{rackname}/{unit-id}

Add or update a host information.

Request example.

Add

    POST /v1/rack/tokyo-0103/22 HTTP/1.1
    Content-Type: application/json
    
    {"name":"www101","cpu":"Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz x 4Core","memory":"16GB","disk":"500GB","os":"ubuntu 16.04","ip":"10.0.100.101/24, 10.0.200.101/24"}

Update

    POST /v1/rack/tokyo-0103/23 HTTP/1.1
    Content-Type: application/json
    
    {"id":1,"name":"www101","cpu":"Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz x 4Core","memory":"16GB","disk":"500GB","os":"ubuntu 16.04","ip":"10.0.100.101/24, 10.0.200.101/24"}

# LICENSE

Copyright (C) ytnobody.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO

https://github.com/ytnobody/kikyo-agent

# AUTHOR

ytnobody <ytnobody@gmail.com>


Changes
Revision history for Perl extension Kikyo

{{$NEXT}}

    - original version



dashboard/gulpfile.js
// generated on 2016-06-07 using generator-webapp 2.1.0
const gulp = require('gulp');
const gulpLoadPlugins = require('gulp-load-plugins');
const browserSync = require('browser-sync');
const del = require('del');
const wiredep = require('wiredep').stream;

const $ = gulpLoadPlugins();
const reload = browserSync.reload;

gulp.task('styles', () => {
  return gulp.src('app/styles/*.scss')
    .pipe($.plumber())
    .pipe($.sourcemaps.init())
    .pipe($.sass.sync({
      outputStyle: 'expanded',
      precision: 10,
      includePaths: ['.']
    }).on('error', $.sass.logError))
    .pipe($.autoprefixer({browsers: ['> 1%', 'last 2 versions', 'Firefox ESR']}))
    .pipe($.sourcemaps.write())
    .pipe(gulp.dest('.tmp/styles'))
    .pipe(reload({stream: true}));
});

gulp.task('scripts', () => {
  return gulp.src('app/scripts/**/*.js')
    .pipe($.plumber())
    .pipe($.sourcemaps.init())
    .pipe($.babel())
    .pipe($.sourcemaps.write('.'))
    .pipe(gulp.dest('.tmp/scripts'))
    .pipe(reload({stream: true}));
});

function lint(files, options) {
  return gulp.src(files)
    .pipe(reload({stream: true, once: true}))
    .pipe($.eslint(options))
    .pipe($.eslint.format())
    .pipe($.if(!browserSync.active, $.eslint.failAfterError()));
}

gulp.task('lint', () => {
  return lint('app/scripts/**/*.js', {
    fix: true
  })
    .pipe(gulp.dest('app/scripts'));
});
gulp.task('lint:test', () => {
  return lint('test/spec/**/*.js', {
    fix: true,
    env: {
      mocha: true
    }
  })
    .pipe(gulp.dest('test/spec/**/*.js'));
});

gulp.task('html', ['styles', 'scripts'], () => {
  return gulp.src('app/*.html')
    .pipe($.useref({searchPath: ['.tmp', 'app', '.']}))
    .pipe($.if('*.js', $.uglify()))
    .pipe($.if('*.css', $.cssnano({safe: true, autoprefixer: false})))
    .pipe($.if('*.html', $.htmlmin({collapseWhitespace: true})))
    .pipe(gulp.dest('dist'));
});

gulp.task('images', () => {
  return gulp.src('app/images/**/*')
    .pipe($.cache($.imagemin({
      progressive: true,
      interlaced: true,
      // don't remove IDs from SVGs, they are often used
      // as hooks for embedding and styling
      svgoPlugins: [{cleanupIDs: false}]
    })))
    .pipe(gulp.dest('dist/images'));
});

gulp.task('fonts', () => {
  return gulp.src(require('main-bower-files')('**/*.{eot,svg,ttf,woff,woff2}', function (err) {})
    .concat('app/fonts/**/*'))
    .pipe(gulp.dest('.tmp/fonts'))
    .pipe(gulp.dest('dist/fonts'));
});

gulp.task('extras', () => {
  return gulp.src([
    'app/*.*',
    '!app/*.html'
  ], {
    dot: true
  }).pipe(gulp.dest('dist'));
});

gulp.task('clean', del.bind(null, ['.tmp', 'dist']));

gulp.task('serve', ['styles', 'scripts', 'fonts'], () => {
  browserSync({
    notify: false,
    port: 9000,
    server: {
      baseDir: ['.tmp', 'app'],
      routes: {
        '/bower_components': 'bower_components'
      }
    }
  });

  gulp.watch([
    'app/*.html',
    'app/images/**/*',
    '.tmp/fonts/**/*'
  ]).on('change', reload);

  gulp.watch('app/styles/**/*.scss', ['styles']);
  gulp.watch('app/scripts/**/*.js', ['scripts']);
  gulp.watch('app/fonts/**/*', ['fonts']);
  gulp.watch('bower.json', ['wiredep', 'fonts']);
});

gulp.task('serve:dist', () => {
  browserSync({
    notify: false,
    port: 9000,
    server: {
      baseDir: ['dist']
    }
  });
});

gulp.task('serve:test', ['scripts'], () => {
  browserSync({
    notify: false,
    port: 9000,
    ui: false,
    server: {
      baseDir: 'test',
      routes: {
        '/scripts': '.tmp/scripts',
        '/bower_components': 'bower_components'
      }
    }
  });

  gulp.watch('app/scripts/**/*.js', ['scripts']);
  gulp.watch('test/spec/**/*.js').on('change', reload);
  gulp.watch('test/spec/**/*.js', ['lint:test']);
});

// inject bower components
gulp.task('wiredep', () => {
  gulp.src('app/styles/*.scss')
    .pipe(wiredep({
      ignorePath: /^(\.\.\/)+/
    }))
    .pipe(gulp.dest('app/styles'));

  gulp.src('app/*.html')
    .pipe(wiredep({
      exclude: ['bootstrap-sass'],
      ignorePath: /^(\.\.\/)*\.\./
    }))
    .pipe(gulp.dest('app'));
});

gulp.task('build', ['lint', 'html', 'images', 'fonts', 'extras'], () => {
  return gulp.src('dist/**/*').pipe($.size({title: 'build', gzip: true}));
});

gulp.task('default', ['clean'], () => {
  gulp.start('build');
});


dashboard/app/scripts/kikyo/rack.js
$(document).ready(function () {
    $(".rackname").text(location.paramObject.rack);
    getRack($(".rack"), location.paramObject.rack);
});

function rackOverview (unitID, row) {
    if (row != null) {
        return $(
            '<tr class="unit" id="unit:'+ unitID +'">'+
              '<td class="unitID">'+unitID+'</td>'+
              '<td class="item name"><a href="/app/input.html?id='+row.id+'">'+row.name+'</a></td>'+
              '<td class="item status">'+row.status+'</td>'+
              '<td class="item virtual_hosts"><a href="/app/search.html?rack='+row.rack+'&unit='+row.unit+'">'+row.virtual_hosts.length+'</a></td>'+
              '<td class="item ip">'+row.ip+'</td>'+
            '</tr>'
        );
    } else {
        return $(
            '<tr class="unit" id="unit:'+ unitID +'">' +
              '<td class="unitID">'+unitID+'</td>' +
              '<td class="name">-</td>' +
              '<td class="status">-</td>' +
              '<td class="virtual_hosts">-</td>' +
              '<td class="ip">-</td>' +
            '</tr>'
        );
    }
}

function getRack (elem, rackname) {
    elem.html("<tr><th>UnitID</th><th>Name</th><th>Status</th><th>Virtual Hosts</th><th>IP</th></tr>");
    $.get("/v1/rack/"+rackname, function(data){
        $.each(data.rows, function(i, val) {
            elem.append(rackOverview(42-i, val));
        });
    });
}


dashboard/app/scripts/kikyo/input.js
$(document).ready(function(){
    if (location.paramObject.id) {
        $.get('/v1/host/'+location.paramObject.id, function(data){
            fillInForm(data.host, $('.hostForm')[0]);
        });
    }
    $('form.hostForm').submit(function(){
        var param = new Object;
        $.each(this.elements, function(i, e){
            if (e.name) {
                if (e.name == "id" && e.value == "") {
                    // nothing.
                }
                else {
                    param[e.name] = e.value;
                }
            }
        });
        if (isValid(param)) {
            registerHost(param);
        }
        return false;
    });
});

function isValid (param) {
    var required = ['rack', 'unit', 'status'];
    var invalid = 0;
    $.each(required, function(i, k){
        if (param[k] == "") {
            invalid = 1;
        }
    });
    return invalid ? false : true;
}

function registerHost (param) {
    var rack = param.rack;
    var unit = param.unit;
    delete param.rack;
    delete param.unit;
    $.ajax({
        type: "POST",
        url: "/v1/rack/"+ rack + "/" + unit,
        contentType: "application/json",
        dataType: "json",
        data: JSON.stringify(param),
        error: function(xhr, stat, err) {
            var errorMessage = xhr.responseJSON.message
            alert('Could not register. '+ errorMessage);
        },
        success: function(data) {
            location.href = "/app/rack.html?rack="+ data.host.rack;
        }
    });
}


dashboard/app/scripts/kikyo/search.js
$(document).ready(function(){
    if (Object.keys(location.paramObject).length > 0) {
        searchHost(location.paramObject, $('.result'));
        fillInForm(location.paramObject, $('.hostSearchForm')[0]);
    }
});

function hostOverview (row) {
    var isVirtual = row.virtual_id > 0 ? '<span class="label label-success">Virtual</span>' : '';
    return $(
        '<tr class="hostOverview">'+
          '<td class="item name"><a href="/app/input.html?id='+row.id+'">'+row.name + isVirtual+'</a></td>'+
          '<td class="item rack">'+row.rack+' / '+row.unit+' / '+row.size+'</td>'+
          '<td class="item hwid">'+row.hwid+'</td>'+
          '<td class="item status">'+row.status+'</td>'+
          '<td class="item ip">'+row.ip+'</td>'+
        '</tr>'
    );
}

function searchHost (cond, elem) {
    elem.html('<tr><th>Name</th><th>Rack / Unit / Size</th><th>HWID</th><th>Status</th><th>IP</th></tr>');
    $.get('/v1/search', cond, function(data) {
        $.each(data.rows, function(i, val){
            elem.append(hostOverview(val));
        });
    });
}



lib/Kikyo.pm
package Kikyo;
use 5.008001;
use strict;
use warnings;

our $VERSION = "0.01";

use Amagi;
use Kikyo::DB;
use Time::Piece;

our $DBC;

sub db () {
    my $time = time;
    if ($DBC && $time - $DBC->[0] > 600) {
        $DBC = undef;
    }
    $DBC ||= [ $time, Kikyo::DB->new ];
    $DBC->[1];
}

get '/' => sub {
    my ($app, $req) = @_;
    $app->redirect('/app/index.html');
};

get '/v1/rack/:rack' => sub {
    my ($app, $req) = @_;
    my $rack = $req->captured->{rack};
    {
        rows => [__PACKAGE__->rackinfo($rack)],
    };
};

get '/v1/host/:id' => sub {
    my ($app, $req) = @_;
    my $host = db->single(host => {id => $req->captured->{id}});
    $host->{virtual_hosts} = [__PACKAGE__->virtual_hosts($host)] if $host->{virtual_id} == 0;
    { host => $host };
};

get '/v1/racklist' => sub {
    my ($app, $req) = @_;
    { 
        rows => [map {$_->{hosts} += 0; $_} db->search_by_sql(
            'SELECT rack, COUNT(unit) AS hosts FROM host WHERE virtual_id=0 GROUP BY rack ORDER BY rack'
        )]
    }
};

get '/v1/search' => sub {
    my ($app, $req) = @_;
    my $query = __PACKAGE__->search_query($req->parameters->mixed);
    { rows => [map {__PACKAGE__->recast_host_fields($_)} db->select(host => $query)] };
};

post '/v1/rack/:rack/:unit' => sub {
    my ($app, $req) = @_;
    my $input = $req->json_content or return $app->res_error(400 => 'Bad Request');
    my $now = localtime->strftime('%Y-%m-%d %H:%M:%S');
    my $id = delete $input->{id};
    my $in_host;
    if ($id) {
        $in_host = {%$input, %{$req->captured}};
        return $app->res_error(400 => 'racking target is collide') if __PACKAGE__->is_corride({%$in_host, id => $id});
        my $old = db->select(host => {id => $id});
        do {
            my $txn = db->txn_scope;
            if ($old->{virtual_id} == 0) {
                db->update(host => 
                    {rack => $in_host->{rack}, unit => $in_host->{unit}},
                    {rack => $old->{rack}, unit => $old->{unit}, virtual_id => {'!=' => 0}}
                );
            }
            db->update(host => $in_host, {id => $id});
            $txn->commit;
        };
    }
    else {
        $in_host = {%$input, %{$req->captured}, create_at => $now};
        return $app->res_error(400, 'racking target is collide') if __PACKAGE__->is_corride($in_host);
        db->insert(host => $in_host);
        $id = db->last_insert_id;
    }
    my $host = db->single(host => {id => $id});
    {host => __PACKAGE__->recast_host_fields($host)};
};

sub search_query {
    my ($class, $params) = @_;
    my $query = +{ map {$_ => $params->{$_}} grep {defined $params->{$_}} qw/id ip name rack hwid os modelname status/ };
    my $key;
    for $key (qw/ip name os modelname/) {
        next unless $query->{$key};
        $query->{$key} = {LIKE => sprintf "%%%s%%", $query->{$key}};
    }
    for $key (qw/rack/) {
        next unless $query->{$key};
        $query->{$key} = {LIKE => sprintf "%s%%", $query->{$key}};
    }
    for $key (keys %$query) {
        delete $query->{$key} if !$query->{$key};
    }
    return $query;
}

sub recast_host_fields {
    my ($class, $host) = @_;
    my $rtn = {%$host};
    $rtn->{$_} = $rtn->{$_} +0 for qw/id unit size/;
    return $rtn;
}

sub host_unit_range {
    my ($class, $host) = @_;
    my $rack = $host->{rack};
    my $unit = $host->{unit};
    my $size = $host->{size} || 1;
    my $unitend = $unit + $size - 1;
    return ($unit .. $unitend);
}

sub is_corride {
    my ($class, $host) = @_;
    my $query = $host->{id} ? 
        {rack => $host->{rack}, virtual_id => $host->{virtual_id}, id => {'!=' => $host->{id}}} : 
        {rack => $host->{rack}, virtual_id => $host->{virtual_id}}
    ;
    my @rows = db->select(host => $query);
    my @filled = map {$class->host_unit_range($_)} @rows;
    my @range = $class->host_unit_range($host);
    for my $i (@range) {
        return 1 if grep {$_ == $i} @filled;
    }
}

sub rackinfo {
    my ($class, $rack) = @_;
    my %item = (map {$_->{unit} => $_} map {$_->{virtual_hosts} = [$class->virtual_hosts($_)]; $_} db->select(host => {rack => $rack, virtual_id => 0}));
    my @rows = map {$item{$_}} 1..42;
    for my $row (grep {defined $_} @rows) {
        my @range = $class->host_unit_range($row);
        $rows[$_-1] = $row for @range;
    }
    reverse @rows;
}

sub virtual_hosts {
    my ($class, $host) = @_;
    my @rows = db->select(host => {rack => $host->{rack}, unit => $host->{unit}, virtual_id => {'!=' => 0}});
    return @rows;
}

1;
__END__

=encoding utf-8

=head1 NAME

Kikyo - Information Base for managing server assets

=head1 DESCRIPTION

Kikyo is a Kikyo is a Information Base for managing server assets.

=head1 HOW TO RUN

=head2 mysql

First, create a database and a user on mysql-server. Then, create table with etc/schema.sql

=head2 environment values

Set following environment values for database connection.

=over 4

=item KIKYO_DSN - dns string for your database

=item KIKYO_DBUSER - database user

=item KIKYO_DBPASS - password for user authentication

=back

Example.

    export KIKYO_DSN="dbi:mysql:host=10.0.0.1;database=kikyo"
    export KIKYO_DBUSER=kikyo
    export KIKYO_DBPASS=kiky0P@ss

Or, modify the app.psgi to conform to your database.

=head2 plackup

Now, you can run Kikyo with plackup under Kikyo project directory.

=head1 JSON API

=head2 get racklist

    GET /v1/racklist

Returns list of rackname and host count.

=head2 get rack info

    GET /v1/rack/{rackname}

Returns hosts list in specified rack.

=head2 fetch host

    GET /v1/host/{hostid}

Returns specified host.

=head2 search hosts

    GET /v1/search

Returns hosts list that matched for search query.

Available search parameters is following.

=over 4

=item id

=item name

=item rack

=item ip

=item hwid

=item os

=item modelname

=item status

=back

=head2 add or update a host

    POST /v1/rack/{rackname}/{unit-id}

Add or update a host information.

Request example.

Add

    POST /v1/rack/tokyo-0103/22 HTTP/1.1
    Content-Type: application/json
    
    {"name":"www101","cpu":"Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz x 4Core","memory":"16GB","disk":"500GB","os":"ubuntu 16.04","ip":"10.0.100.101/24, 10.0.200.101/24"}


Update

    POST /v1/rack/tokyo-0103/23 HTTP/1.1
    Content-Type: application/json
    
    {"id":1,"name":"www101","cpu":"Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz x 4Core","memory":"16GB","disk":"500GB","os":"ubuntu 16.04","ip":"10.0.100.101/24, 10.0.200.101/24"}


=head1 LICENSE

Copyright (C) ytnobody.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

https://github.com/ytnobody/kikyo-agent

=head1 AUTHOR

ytnobody E<lt>ytnobody@gmail.comE<gt>

=cut

